examples/phase_change/ex0/SetFluidProperties.cpp

```diff
@@ -0,0 +1,345 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2017 - 2019 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+// APPLICATION INCLUDES
+#include "SetFluidProperties.h"
+
+#include <ibamr/app_namespaces.h>
+
+// C++ INCLUDES
+
+/////////////////////////////// STATIC ///////////////////////////////////////
+void
+callSetLiquidSolidDensityCallbackFunction(int rho_idx,
+                                          Pointer<Variable<NDIM> > rho_var,
+                                          Pointer<IBTK::HierarchyMathOps> hier_math_ops,
+                                          const int cycle_num,
+                                          const double time,
+                                          const double current_time,
+                                          const double new_time,
+                                          void* ctx)
+{
+    // Set the density from the level set information
+    static SetFluidProperties* ptr_SetFluidProperties = static_cast<SetFluidProperties*>(ctx);
+    ptr_SetFluidProperties->setDensityPatchData(
+        rho_idx, rho_var, hier_math_ops, cycle_num, time, current_time, new_time);
+
+    return;
+
+} // callSetFluidGasSolidDensityCallBackFunction
+
+void
+callSetLiquidSolidSpecificHeatCallbackFunction(int Cp_idx,
+                                               Pointer<Variable<NDIM> > Cp_var,
+                                               Pointer<IBTK::HierarchyMathOps> hier_math_ops,
+                                               const int cycle_num,
+                                               const double time,
+                                               const double current_time,
+                                               const double new_time,
+                                               void* ctx)
+{
+    // Set the density from the level set information
+    static SetFluidProperties* ptr_SetFluidProperties = static_cast<SetFluidProperties*>(ctx);
```

Comment:
The main idea behind this design was that we would like to set density/viscosity using any sort of interface capturing method (level set, VoF, phase field), and the user can define it based upon their choice during the application setup. We done this using level set and phase field as of now, but in the future we will also like to set it using VoF.

Reply:
Perhaps `SetLSProperties.(h,cpp)` can be a helper function in the library. `GravityForcing.(cpp, h)` can also be, but the idea of Gravity forcing is that a user should be able to define it in any suitable manner (depending upon the interface tracking method). `TagInterfaceRefinementCells.(h,cpp)` cannot be because it is application dependent (what to tag for refinement).

Reply:
I'd like to keep application-specific stuff out of the library itself.

It might make sense to have a helper library for the examples to avoid code duplication.

On the other hand, I have often thought of the examples as providing self-contained starter code for a "real" application -- in which case it is helpful to have everything in place in the example directory.

Reply:
> 
> On the other hand, I have often thought of the examples as providing self-contained starter code for a "real" application -- in which case it is helpful to have everything in place in the example directory.

Yes, that is the idea behind the examples. It is just that we have many (useful) examples that are benchmarking cases from the literature.